tools/fork-network/src/cli.rs

```diff
@@ -1,191 +1,662 @@
-                        if hash != flat_head_for_this_shard {
-                            panic!("Not all shards have the same flat head. Please reset the fork, and run the node for a little longer");
-                        }
-                    }
-                    flat_head = Some(flat_head_for_this_shard);
-                } else {
-                    panic!(
-                        "Flat storage is not ready for shard {}: {:?}. Please reset the fork, and run the node for longer",
-                        shard_uid, flat_storage_status
-                    );
-                }
-            }
-            println!("Forking from the flat storage final head: {}", flat_head.unwrap());
-            flat_head.unwrap()
-        };
-
-        let fork_head_block = store
-            .get_ser::<near_primitives::block::Block>(
-                DBCol::Block,
-                &borsh::to_vec(&fork_head).unwrap(),
-            )?
-            .unwrap();
+        let fork_head = get_fork_head(&all_shard_uids, store.clone())?;
+        tracing::info!(?fork_head);
+        let fork_head_block =
+            store.get_ser::<Block>(DBCol::Block, &borsh::to_vec(&fork_head).unwrap())?.unwrap();
 
+        let runtime =
+            NightshadeRuntime::from_config(home_dir, store.clone(), &near_config, epoch_manager);
         let prev_state_roots = fork_head_block
             .chunks()
             .iter()
             .map(|chunk| chunk.prev_state_root())
             .collect::<Vec<_>>();
 
-        // TODO(robin-near): Make this part customizable via some kind of config file.
-        let self_validator = near_config.validator_signer.as_ref().unwrap();
-        let self_account = AccountInfo {
-            account_id: self_validator.validator_id().clone(),
-            amount: 50_000 * NEAR_BASE,
-            public_key: self_validator.public_key(),
-        };
+        let prev_hash = *fork_head_block.header().prev_hash();
+        let make_storage_mutator: MakeSingleShardStorageMutatorFn =
+            Arc::new(move |shard_id, prev_state_root| {
+                SingleShardStorageMutator::new(
+                    shard_id,
+                    &runtime.clone(),
+                    prev_hash,
+                    prev_state_root,
+                )
+            });
 
-        let mut storage_mutator = StorageMutator::new(
-            epoch_manager.clone(),
-            &*runtime,
-            fork_head_block.header().epoch_id(),
-            *fork_head_block.header().prev_hash(),
+        let new_state_roots = self.prepare_state(
+            batch_size,
+            &all_shard_uids,
+            store,
             &prev_state_roots,
+            make_storage_mutator.clone(),
         )?;
+        Ok(new_state_roots)
+    }
+
+    /// Creates a DB snapshot, then
+    /// Reads a list of validator accounts from a file
+    /// Adds validator accounts to the state
+    /// Creates a genesis file with the new validators.
+    fn set_validators(
```

Comment:
when I tried running this one after running `amend-access-keys`, I got: `called `Result::unwrap()` on an `Err` value: MissingTrieValue(TrieStorage, Aw8vwhWtg96GNd7j4cuehERSo2tDiaRWmhsYsSxfEwZD)` so there is somethign going wrong somewhere, but i didnt investigate further. i still have the home dirs so i can do some more debugging later if it helps

Reply:
Thanks for testing. This is a big problem.
Solved it by persisting state roots in the `Misc` column.